### PR TITLE
fix hipStreamAddCallback race

### PR DESCRIPTION
**Pull request by jeffdaily**

### CODE REVIEW

**Comment by srinivamd** [comment]
Does this address this issue https://github.com/ROCm-Developer-Tools/HIP/issues/1205 also?

**Comment by jeffdaily** [comment]
I wasn't aware of #1205.  I will check.  I know that this PR does fix a few TensorFlow unit tests, and that's why it was written.

**Comment by jeffdaily** [comment]
I have 1 assert left from the test code attached to #1205 .
```
===============================================================================
test cases: 10 |  9 passed | 1 failed
assertions: 20 | 19 passed | 1 failed
```

The last failure might be due to HCC.  This PR is the first that I am aware of that uses `accelerator_view::acquire_locked_hsa_queue` to insert foreign AQL packets.  This is causing some failures in HCC.

**Comment by srinivamd** [comment]
@jeffdaily Your results, with just 1 assertion failed, is awesome! Will start reviewing the code.

File: src/hip_hcc.cpp [hunk 1]
```diff
@@ -306,17 +318,21 @@ inline void ihipStream_t::locked_wait(bool& waited) {
     // create a marker while holding stream lock,
     // but release lock prior to waiting on the marker
     hc::completion_future marker;
+    std::vector<hsa_signal_t> pending_callbacks;
```
Comment (srinivamd) [inline src/hip_hcc.cpp]: initialize it `{}` just to be safe?

File: src/hip_stream.cpp [hunk 2]
```diff
@@ -257,11 +257,85 @@ hipError_t hipStreamGetPriority(hipStream_t stream, int* priority) {
...
+    // stream is locked, now lock underlying HSA queue using unique_ptr RAII
+    auto av{cs->_av};
+    auto d{[=](hsa_queue_t*) mutable { av.release_locked_hsa_queue(); }};
+    std::unique_ptr<hsa_queue_t, decltype(d)> q{
+        static_cast<hsa_queue_t*>(av.acquire_locked_hsa_queue()),
+        std::move(d)};
+
+    // get first packet
+    auto b_index{hsa_queue_add_write_index_relaxed(q.get(), 1) % q->size};
+    auto b{static_cast<hsa_barrier_or_packet_t*>(q->base_address) + b_index};
```
Comment (srinivamd) [inline src/hip_stream.cpp]: nit: not sure about single letter variable identifiers 'b' and 'd'. Maybe a more meaningful/longer identifier, making it easier to spot it in code, improve readability?

**Comment by jeffdaily** [comment]
@srinivamd I appreciate your review.  I would make the suggested changes, but we need to have more discussion with HCC team concerning this PR.

**Comment by jeffdaily** [comment]
Closing in favor of #1921 .  This current PR cannot be accepted since it breaks a contract with HCC's runtime since HCC cannot track dependencies of foreign AQL packets.

**Pull request closed without merging**